publication-commons/src/main/java/no/unit/nva/publication/utils/CristinUnitsUtil.java

```diff
@@ -0,0 +1,146 @@
+    private static final Map<String, CristinUnit> allUnits = new HashMap<>();
+
+    public static final String APPLICATION_JSON = "application/json";
+    private final URI apiUri;
+    private final UriRetriever uriRetriever;
+
+    public CristinUnitsUtil(UriRetriever uriRetriever, URI apiUri) {
+        this.uriRetriever = uriRetriever;
+        this.apiUri = apiUri;
+    }
+
+    /***
+     * Get the top level unit for a given unit.
+     * @param unitId URI of the unit
+     * @return URI of the top level unit or null if not found
+     */
+    public URI getTopLevel(URI unitId) {
+        if (allUnits.isEmpty()) {
+            loadAllData();
+            buildTree();
+            logger.info("Loaded cristin unit count: {}", allUnits.size());
+        }
+
+        var lookupId = UriWrapper.fromUri(unitId).getLastPathElement();
+
+        return findTopLevelUnitOrSelf(lookupId).map(a -> URI.create(unitId.toString().replace(lookupId, a.id())))
+                   .orElse(null);
+    }
+
+    private static void buildTree() {
+        for (var unit : allUnits.values()) {
+            if (nonNull(unit.parentUnit())) {
+                var parent = allUnits.get(unit.parentUnit().id());
+                if (parent.children() == null) {
+                    parent = new CristinUnit(parent.id(), new ArrayList<>(), parent.parentUnit());
+                }
+                parent.children().add(unit);
+            }
+        }
+    }
+
+    private void loadAllData() {
+        int pageNum = 1;
+
+        while (true) {
+            var units = getApiData(pageNum);
+            if (units.length == 0) {
+                break;
+            } else {
+                for (var unit : units) {
+                    allUnits.put(unit.id(), unit);
+                }
+                pageNum++;
+            }
+        }
+    }
+
+    private static Optional<CristinUnit> findTopLevelUnitOrSelf(String unitId) {
+        var unit = CristinUnitsUtil.allUnits.get(unitId);
+        if (unit == null) {
+            return Optional.empty();
+        }
+        while (unit.parentUnit() != null) {
+            unit = CristinUnitsUtil.allUnits.get(unit.parentUnit().id());
+        }
+        return Optional.of(unit);
+    }
+
+    private CristinUnit[] getApiData(int pageNum) {
+        try {
+            String response = fetchResponseWithRetry(new URI(apiUri + API_ARGUMENTS + pageNum));
+            return attempt(() -> JsonUtils.dtoObjectMapper.readValue(response, CristinUnit[].class)).orElseThrow();
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private String fetchResponseWithRetry(URI requestUri) {
+        var retryRegistry = RetryRegistry.of(RetryConfig.custom().maxAttempts(5).intervalFunction(
+            IntervalFunction.ofExponentialRandomBackoff()).build());
+        var retryWithDefaultConfig = retryRegistry.retry("executeRequestAsync");
```

Comment:
`executeRequestAsync `?? your supplier is `executeRequest `?

Reply:
https://github.com/BIBSYSDEV/nva-cristin-service/blob/main/cristin-commons/src/test/java/no/unit/nva/cristin/common/client/ApiClientTest.java

https://github.com/BIBSYSDEV/nva-cristin-service/blob/main/cristin-commons/src/main/java/no/unit/nva/client/RetryConfigProvider.java

Reply:
If the supplier is named `executeRequest `this should also be applied to registry: `retryRegistry.retry("executeRequest")`

Reply:
Maybe write a test for it

Reply:
copy pasta bug. did a unit test and deleted it afterwards as I don't see any point keeping it :-)